test/integration/controlplane/kms_transformation_test.go

```diff
@@ -176,14 +178,101 @@ resources:
+		t.Fatalf("failed to encrypt future secret, err: %v", err)
+	}
+
+	futureEncryptedSecretBuf := cryptobyte.NewBuilder(nil)
+	futureEncryptedSecretBuf.AddBytes([]byte(wantPrefix))
+	futureEncryptedSecretBuf.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes([]byte(futureAESGCMKey))
+	})
+	futureEncryptedSecretBuf.AddBytes(futureEncryptedSecretBytes)
+
+	_, err = test.writeRawRecordToETCD(secretETCDPath, futureEncryptedSecretBuf.BytesOrPanic())
+	if err != nil {
+		t.Fatalf("failed to write future encrypted secret, err: %v", err)
+	}
+
+	// confirm that direct AES CBC decryption does not work
+	failingRawEnvelope, err := test.getRawSecretFromETCD()
+	if err != nil {
+		t.Fatalf("failed to read %s from etcd: %v", secretETCDPath, err)
+	}
+	failingFutureEnvelope := envelope{
+		providerName: providerName,
+		rawEnvelope:  failingRawEnvelope,
+		plainTextDEK: futureKeyBytes,
+	}
+	failingFuturePlainSecret, err := failingFutureEnvelope.plainTextPayload(secretETCDPath)
+	if err == nil || err.Error() != "failed to transform from storage via AESCBC, err: the stored data is not a multiple of the block size" {
+		t.Fatalf("AESCBC decryption failure not seen, err: %v, data: %s", err, string(failingFuturePlainSecret))
+	}
+
+	// AES GCM secrets should be un-enveloped on direct reads from Kube API Server.
+	futureSecretObj, err := secretClient.Get(context.TODO(), testSecret, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to read future secret via Kube API, err: %v", err)
+	}
+	if futureSecretVal != string(futureSecretObj.Data[secretKey]) {
+		t.Fatalf("expected %s from KubeAPI, but got %s", futureSecretVal, string(futureSecretObj.Data[secretKey]))
+	}
+
+	// no-op update should cause new AES CBC key to be used
+	futureSecretUpdated, err := secretClient.Update(context.TODO(), futureSecretObj, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("failed to update future secret via Kube API, err: %v", err)
+	}
+	if futureSecretObj.ResourceVersion == futureSecretUpdated.ResourceVersion {
+		t.Fatalf("future secret not updated on no-op write: %s", futureSecretObj.ResourceVersion)
+	}
+
+	// confirm that direct AES CBC decryption works
+	futureRawEnvelope, err := test.getRawSecretFromETCD()
+	if err != nil {
+		t.Fatalf("failed to read %s from etcd: %v", secretETCDPath, err)
+	}
+	futureEnvelope := envelope{
+		providerName: providerName,
+		rawEnvelope:  futureRawEnvelope,
+		plainTextDEK: pluginMock.LastEncryptRequest(),
+	}
+	if !bytes.HasPrefix(futureRawEnvelope, []byte(wantPrefix)) {
+		t.Fatalf("expected secret to be prefixed with %s, but got %s", wantPrefix, futureRawEnvelope)
+	}
+	futurePlainSecret, err := futureEnvelope.plainTextPayload(secretETCDPath)
+	if err != nil {
+		t.Fatalf("failed to transform from storage via AESCBC, err: %v", err)
+	}
+	if !strings.Contains(string(futurePlainSecret), futureSecretVal) {
+		t.Fatalf("expected %q after decryption, but got %q", futureSecretVal, string(futurePlainSecret))
+	}
```

Comment:
Nit: quite a lot of things are going on. Would it make sense to break it up, even though this might be just a temporary test case?

Reply:
We're only doing a string check. Do you have suggestion on how we should break this up?

Reply:
Oh, sorry. I was not specific enough. I meant the whole test case, not the string check itself.